### ISSUE TITLE
Better Validation-Accuracy with Tensorflow than MxNet 

### ISSUE THREAD

**Issue by SaharIr** [apache/incubator-mxnet#5760]
I tried to compare the speed and the validation accuracy of  a Convolutional Neural Network with 5 layers (3 convolutional layers and 2 fully connected layers) on MxNet and Tensorflow. For MNIST dataset , It shows that with the almost same accuracy, MxNet is quite faster than Tensorflow. However, I did this comparison with Dogs vs. Cats. It shows that MxNet cannot be as good as Tensorflow in accuracy. 
I have also tried Residual blocks, and again It demonstrated that the accuracy is not comparable with the accuracy that can be achieved by Tensorflow.

 I used mx.model.FeedForward.create for training, and mx.io.NDArrayIter for data traitors. I repeated the experiment with different sgd optimizer like Adam, sgd+momentom and RMSProp. However, couldn't achieve the same accuracy as Tensorflow.

Could you please help me to understand that where the problem might be from. Is it from Mxnet functions that I used or in general MxNet in not as accurate as Tensorflow?
Thank you in advance for your consideration.

**Comment by zhreshold** [comment]
Can you provide the example?  Details matter, such as data augmentation.
I am also curious to see how these performance gaps could emerge.

**Comment by SaharIr** [comment]
CNN example: 
      - Image size=64x64 (without data augmentation) 
      - 3 Convolutional layers (num_filter= 16(first layer), 32(second layer), 64(third layer) and kernel=(3,3) for all ) +  fully connected layer(num_hidden=10)
      - SGD ( momentum )
      - mx.io.NDArrayIter for data iterator(  batch_size=50)
      - Training: model = mx.model.FeedForward.create(    ctx = mx.gpu(),   symbol=network,   num_epoch = 100,    X=train_iter, eval_data=val_iter,  learning_rate=0.0005, momentum=0.9 )

I tested the same network in Tensorflow, and these are the results:
                                                                                                  MxNet:       Validation-Accuracy=68.01 %
                                                                                                  Tensorflow: Validation Accuracy=74.06%

ResNet: 
      - Image size=64x64 (without data augmentation)  
      - 36 layers (1Convolution layer + 15 Residual-block + 1 fully connected layer).
      -  Residual-block (  BatchNorm+Activatin+Convolution+BatchNorm+Activation+Convolution). To increase the channels of Residual input to be as same size as its output I used the Convolution with kernel=(1,1).  The Mxnet Padding function in not working for 0 and 1 dimensions.
      - - mx.io.NDArrayIter for data iterator(  batch_size=50)
       - Training:  model = mx.model.FeedForward.create(  ctx = mx.gpu(),  symbol=network, num_epoch = 100,  optimizer = mx.optimizer.Adam(),  X=train_iter, eval_data=val_iter,  learning_rate=0.0005, momentum=0.9)

I also tested the same network in Tensorflow, and these are the results:
                                                                                                  MxNet:       Validation-Accuracy=71.45 %
                                                                                                  Tensorflow: Validation-Accuracy=82.5%
Tensorflow also shows  much more better performances with other configurations like larger image size and batch size. However Mxnet is saturated to 71%.

**Comment by zhreshold** [comment]
What about the Initializer?

**Comment by Jerryzcn** [comment]
Did you turn off the data augmentation for val_iter? Also, you want to scale all your validation data to 256x256 (or whatever that you use for min_random_scale).

**Comment by SaharIr** [comment]
mx.initializer.Xavier(), set  factor_type="in", magnitude=2.34, and also without these parameters.

**Comment by SaharIr** [comment]
I didn't use data augmentation.  mx.io.NDArrayIter doesn’t have augmentation parameter

**Comment by zhreshold** [comment]
@SaharIr  Are you using Xavier init for tensorflow as well, and the params in tensorflow initializer?.

**Comment by SaharIr** [comment]
@zhreshold yes. I used same initializer and same params.

**Comment by jmerkow** [comment]
Can you post the actual code for both, perhaps in a gist?  This data is your data? or a public dataset?

**Comment by SaharIr** [comment]
@jmerkow I'm using Dogs vs. Cats from kaggle.

Please find the following links including CNN example codes in MxNet and Tensorflow: 

https://gist.github.com/SaharIr/fb0fe482bd9d62d64a2aaa4d5f92f4f7

https://gist.github.com/SaharIr/67cdc9ab4ed4f4d9020c8f1a2bcfac79

**Comment by SaharIr** [comment]
Is there any update about my problem? I would appreciate if someone can help me to find the reason why Tensorflow has better performances than MxNet ? Is there any thing wrong with my codes or the problem is comming from MxNet?

**Comment by zhreshold** [comment]
For mxnet, line 85
```
mx.initializer.Xavier()
```
This is useless, you must specify it in the create function
```
model = mx.model.FeedForward.create(
    ctx = mx.gpu(),    
    symbol=NN_model,       
    num_epoch = 100,     
    X=nd_iter, 
    eval_data=val_iter,
    learning_rate=0.0005, momentum=0.9,
    initializer=mx.initializer.Xavier()
)
```
Otherwise it's using default Uniform initializer

**Comment by SaharIr** [comment]
I've applied that on my code, but it haven't solved the problem. Is there anything else that I could apply to merge these performances?

**Comment by zhreshold** [comment]
I look at you mxnet code again, change
```
allX = allX.reshape(allX.shape[0],3,64,64)
```
to 
```
allX = allX.transpose(0, 3, 1, 2)
```
You are directly reshaping so the images are corrupted along x axis

**Comment by SaharIr** [comment]
Thank you very much for your taking time. I change the code based on your comment and it is working! I got the same accuracy as Tensorflow.

**Comment by szha** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to ping me to reopen if this is still an active issue. Thanks!
